caffe2/operators/crop_and_resize_op.cc

```diff
@@ -0,0 +1,366 @@
+
+                        const int nearest_x_index = _roundf(in_x);
+                        const int nearest_y_index = _roundf(in_y);
+
+                        crops_data[b * channels * crop_height_ * crop_width_ +
+                            c * crop_height_ * crop_width_ +
+                            y * crop_width_ +
+                            x] = fm_data[batch_index * channels * fm_height * fm_width +
+                                    c * fm_height * fm_width +
+                                    nearest_y_index * fm_width +
+                                    nearest_x_index];
+                    }
+
+                }
+            }
+        }
+    }
+
+    return true;
+}
+
+template<>
+bool CropAndResizeGradientOp<float, CPUContext>::RunOnDevice() {
+
+    const auto& X = Input(0);
+    const auto& boxes = Input(1);
+    const auto& dY = Input(2);
+
+    auto* dX = Output(0);
+
+    dX->ResizeLike(X);
+
+    CAFFE_ENFORCE_EQ(dX->dim(), 4);
+    CAFFE_ENFORCE_EQ(boxes.dim(), 2);
+
+    // [batch, x1, y1, x2, y2]
+    CAFFE_ENFORCE_EQ(boxes.dim32(1), 5);
+
+    CAFFE_ENFORCE_EQ(dY.dim32(0), boxes.dim32(0));
+    CAFFE_ENFORCE_EQ(dY.dim32(1), dX->dim32(1));
+    CAFFE_ENFORCE_EQ(dY.dim32(2), crop_height_);
+    CAFFE_ENFORCE_EQ(dY.dim32(3), crop_width_);
+
+    const int batch_size = dX->dim32(0);
+    const int channels = dX->dim32(1);
+    const int fm_height = dX->dim32(2);
+    const int fm_width = dX->dim32(3);
+
+    const auto* X_data = X.template data<float>();
+    const auto* boxes_data = boxes.template data<float>();
+    const auto* dY_data = dY.template data<float>();
+
+    auto* dX_data = dX->template mutable_data<float>();
+
+    math::Set<float, CPUContext>(
+        dX->numel(), 0.0f, dX_data, &context_);
+
+    const int box_dim = boxes.dim32(1);
+
+    for (int b = 0; b < boxes.dim32(0); ++b) {
+
+        const int batch_index = static_cast<int>(boxes_data[b * box_dim]);
+
+        if (batch_index < 0 || batch_index >= batch_size) {
+            continue;
+        }
+
+        const float x1 = boxes_data[b * box_dim + 1];
+        const float y1 = boxes_data[b * box_dim + 2];
+        const float x2 = boxes_data[b * box_dim + 3];
+        const float y2 = boxes_data[b * box_dim + 4];
+
+        const float height_scale = (crop_height_ > 1) ?
+            ((y2 - y1) * (fm_height - 1) / (crop_height_ - 1)) : 0.0f;
+
+        const float width_scale = (crop_width_ > 1) ?
+            ((x2 - x1) * (fm_width - 1) / (crop_width_ - 1)) : 0.0f;
+
+        for (int c = 0; c < channels; ++c) {
+            for (int y = 0; y < crop_height_; ++y) {
+
+                const float in_y = (crop_height_ > 1) ?
+                    (y1 * (fm_height - 1) + y * height_scale) :
+                    (0.5f * (y1 + y2) * (fm_height - 1));
+
+
+                if (in_y < 0 || in_y > fm_height - 1) {
+                    continue;
+                }
+
+                if (method_ == std::string("bilinear")) {
```

Comment:
It's fine here, but I'd suggest to make method_ as an instance of a enum class.

Reply:
Switched it to regular `enum` if it's ok.
The reason why not exactly `enum class` is because there seems to be a need to overload output operators for `enum class` for log checks.